### PR TITLE
Support for Path objects in io packages

### PR DESCRIPTION
**Pull request by anchitjain1234**
For #4412.
Right now added support for ``io.ascii`` by modifying ``get_readable_fileobj`` in ``utils`` because ``io.ascii`` uses ``get_readable_fileobj`` to read.
I would also try to add support for Path objects in other io packages.

### CODE REVIEW

**Comment by anchitjain1234** [comment]
@taldcroft Is this the correct approach? I tried to follow yours as in #4405.

**Comment by taldcroft** [comment]
@anchitjain1234 - thanks, this looks promising!  If you do a census throughout the project of places that use `get_readable_fileobj`, I would be nice to see demonstration through testing of this fix being more broadly applicable than just `io.ascii`.  I'm not sure off-hand how `io.fits` and `io.votable` handle the front-end task of handling the "input" file, which might be one of many different object types.

**Comment by anchitjain1234** [comment]
> If you do a census throughout the project of places that use get_readable_fileobj, I would be nice to see demonstration through testing of this fix being more broadly applicable than just io.ascii.

@taldcroft I grepped the whole project to find the usage of ``get_readable_fileobj``. Check results [here](https://gist.github.com/anchitjain1234/154b993e4614b7e8a791).
From the above results, I think I have covered through tests ``io.ascii``, ``io.votable`` and ``utils``.
But I am not sure if ``vo`` also needs to be tested?

> I'm not sure off-hand how io.fits and io.votable handle the front-end task of handling the "input" file

I handled ``io.fits`` directly in ``_File`` class and ``io.votable`` is automatically accetping ``pathlib.Path object`` because it uses ``utils.xml.iterparser`` which in turn uses ``get_readable_fileobj``.

I also added a separate test for ``votable.validate`` because it uses ``get_readable_fileobj`` directly to get content.

**Comment by anchitjain1234** [comment]
@taldcroft Could you please check the changes?

File: astropy/io/fits/file.py [hunk 1]
```diff
@@ -97,6 +104,9 @@ def __init__(self, fileobj=None, mode=None, memmap=None, clobber=False,
             return
         else:
             self.simulateonly = False
+            # If fileobj is of type pathlib.Path
+            if PY3 and HAS_PATHLIB and isinstance(fileobj, Path):
```
Comment (taldcroft) [inline astropy/io/fits/file.py]: You don't need to check `PY3`, all that matters is `HAS_PATHLIB` to make sure the second logical test will execute properly.

File: astropy/io/fits/file.py [hunk 2]
```diff
@@ -74,6 +74,13 @@
 PKZIP_MAGIC = b('\x50\x4b\x03\x04')
 BZIP2_MAGIC = b('\x42\x5a')
 
+try:
+    from pathlib import Path
```
Comment (taldcroft) [inline astropy/io/fits/file.py]: Make this `import pathlib` for consistency with usage in the other modules.

File: astropy/io/votable/tests/vo_test.py [hunk 3]
```diff
@@ -820,6 +827,45 @@ def test_validate():
     assert truth == output
 
 
+@pytest.mark.skipif('not HAS_PATHLIB')
+def test_validate_path_object():
+    """
+    Validating when source is passed as path object. (#4412)
+    Creating different method from ``test_validate`` so that it could be
```
Comment (taldcroft) [inline astropy/io/votable/tests/vo_test.py]: I think you can just add an argument to `test_validate`, e.g.
```
def test_validate(test_path_object=False):
    fpath = get_pkg_data_filename('data/regression.xml')
    if test_path_object:
        fpath = pathlib.Path(fpath)
    ...
```
Then you can still have a separate `test_validate_path_object()` (which will be skipped for PY2) but just calls `test_validate(test_path_object=True)`.

File: astropy/utils/data.py [hunk 4]
```diff
@@ -167,6 +167,14 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
     # passed in.  In that case it is not the responsibility of this
     # function to close it: doing so could result in a "double close"
     # and an "invalid file descriptor" exception.
+    PATH_TYPES = six.string_types
+    try:
+        from pathlib import Path
```
Comment (taldcroft) [inline astropy/utils/data.py]: Better to follow the same idiom from testing and define `HAS_PATHLIB`.  This allows for the possibility of a Python 4 that may or may not have pathlib.

Note: you should change my original implementation in `io.registry` to match (#4405).  I fell into the mistake of testing for `PY3`.  See http://astrofrog.github.io/blog/2016/01/12/stop-writing-python-4-incompatible-code/.  In fact the fix to `io.registry` might not be needed now that you are patching `get_readable_fileobj`.  I haven't looked through the logic that closely but you should look and test.

File: astropy/utils/data.py [hunk 5]
```diff
@@ -175,7 +183,11 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
         remote_timeout = conf.remote_timeout
 
     # Get a file object to the content
-    if isinstance(name_or_obj, six.string_types):
+    if isinstance(name_or_obj, PATH_TYPES):
+        # name_or_obj could be a Path object in Python 3
+        if six.PY3:
```
Comment (taldcroft) [inline astropy/utils/data.py]: Notice that this code becomes easier to understand:
```
        if HAS_PATHLIB:
            ...
```
This avoids one level of contextual knowledge that only Python 3 has pathlib.  So then the comment would be more like `# name_or_obj could be a Path object if pathlib is available`.

**Comment by taldcroft** [comment]
@anchitjain1234 - looks pretty good, thanks!

**Comment by saimn** [comment]
Also pathlib can be installed on PY2: https://pypi.python.org/pypi/pathlib2/
So it is better to avoid testing for PY3 as @taldcroft said.

**Comment by anchitjain1234** [comment]
@taldcroft Please check the changes.

File: astropy/io/registry.py [hunk 6]
```diff
@@ -321,9 +322,9 @@ def read(cls, *args, **kwargs):
             if len(args):
                 if isinstance(args[0], PATH_TYPES):
                     from ..utils.data import get_readable_fileobj
-                    # For Py3 the path might be a pathlib.Path object, so coerce
-                    # to a regular string.
-                    if six.PY3:
+                    # path might be a pathlib.Path object if HAS_PATHLIB,
+                    # so coerce to a regular string.
+                    if HAS_PATHLIB:
```
Comment (taldcroft) [inline astropy/io/registry.py]: It occurs to me that in theory a user might be passing a unicode file name on Python 2 with the backport of pathlib independently installed.  Then conversion with `str()` would be a problem.  So maybe you need to also check `isinstance(args[0], pathlib.Path)`?

Comment (anchitjain1234) [inline astropy/io/registry.py]: @taldcroft If I understand correctly, you meant to say for example a file named ``x©.fits`` on Python 2 with the backport of pathlib independently installed would be a problem?
I cant understand why would it be a problem? 
I also tested it on my system on ``Python 2 `` and ``Python 3`` where it works correctly.

**Comment by taldcroft** [comment]
@anchitjain1234 - looks good!  Now you just need to add an entry to CHANGES.rst describing this update.  You can put `[skip ci]` in the commit message to avoid re-running CI testing.

**Comment by anchitjain1234** [comment]
@taldcroft In ``Python 2 `` pathlib is ``pathlib 2`` therefore ``import pathlib`` would not work. So do I need to add ``import pathlib2`` statements also?

**Comment by saimn** [comment]
@anchitjain1234 - Even if the package name is pathlib2, it will import as `pathlib`.

@taldcroft - Maybe the extra dependency should be added to a Python2 build in Travis ?

**Comment by anchitjain1234** [comment]
@taldcroft Changelog entry added.

**Comment by taldcroft** [comment]
@saimn - I think that @anchitjain1234 has a point.  If pathlib2 is installed the code still needs to explicitly do:
```
import pathlib2 as pathlib
```
I was hoping that `pip install pathlib2` did a magic rename so that `import pathlib` would work on Py2, but I just tried and that is not the case.   So this leads to the uglyness:
```
try:
   import pathlib
except ImportError:
    try:
        import pathlib2 as pathlib
.. etc
```
I think we should just merge this current version, which addresses the stated issue of providing support for bona-fide `pathlib.Path` objects.  Supporting the Python2 backport can be a separate issue, but I think it will need to go along with refactoring the `HAS_PATHLIB` logic into a separate single entity.

**Comment by saimn** [comment]
@taldcroft - Ah sorry, I tested this on a virtualenv, that I thought was based on PY2 but it seems not. The backport provides indeed a `pathlib2` package, and this is intentional : https://github.com/mcmtroffaes/pathlib2/issues/8 (Probably for good reasons even if it is not handy).

**Comment by taldcroft** [comment]
Thanks @anchitjain1234 !  And @saimn for useful review and input.

**Pull request merged**